### PR TITLE
Update ButterflyPACK and STRUMPACK

### PR DESCRIPTION
**Pull request by pghysels**
Add new versions of ButterflyPACK (to 1.1.0) and STRUMPACK (to 3.3.0, and 3.2.0).
Add a dependency from STRUMPACK to ButterflyPACK.

@liuyangzhuan 
@balay

### CODE REVIEW

File: var/spack/repos/builtin/packages/butterflypack/package.py [hunk 1]
```diff
@@ -22,13 +22,14 @@ class Butterflypack(CMakePackage):
...
     git      = "https://github.com/liuyangzhuan/ButterflyPACK.git"
-    url      = "https://github.com/liuyangzhuan/ButterflyPACK/archive/v1.0.3.tar.gz"
+    url      = "https://github.com/liuyangzhuan/ButterflyPACK/archive/v1.1.0.tar.gz"
     maintainers = ['liuyangzhuan']
 
     version('master', branch='master')
     version('1.0.0', sha256='86c5eb09a18522367d63ce2bacf67ca1c9813ef351a1443baaab3c53f0d77232')
     version('1.0.1', sha256='e8ada37466a19f49e13456b150700d4c3afaad2ddbe3678f4e933f9d556a24a5')
     version('1.0.3', sha256='acf9bc98dd7fea31ab73756b68b3333228b53ab0e85400a8250fcc749a1a6656')
+    version('1.1.0', sha256='0e6fd0f9e27b3ee8a273dc52f4d24b8737e7279dc26d461ef5658b317215f1dc')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/butterflypack/package.py]: Can you sort these from newest to oldest?

File: var/spack/repos/builtin/packages/strumpack/package.py [hunk 2]
```diff
@@ -59,8 +63,11 @@ class Strumpack(CMakePackage):
     depends_on('parmetis', when='+parmetis')
     depends_on('scotch~metis', when='+scotch')
     depends_on('scotch~metis+mpi', when='+scotch+mpi')
+    depends_on('butterflypack@1.1.0', when='+butterflypack+mpi')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/strumpack/package.py]: Is this the only version that will work?

File: var/spack/repos/builtin/packages/strumpack/package.py [hunk 3]
```diff
@@ -90,11 +97,17 @@ def on_off(varstr):
...
+                '-DTPL_ENABLE_BPACK=%s' % on_off('+butterflypack')
             ])
         else:
             args.extend([
                 '-DSTRUMPACK_USE_PARMETIS=%s' % on_off('+parmetis'),
                 '-DSTRUMPACK_USE_SCOTCH=%s' % on_off('+scotch')
             ])
+
+        args.extend([
+            '-DBUILD_SHARED_LIBS=%s' % on_off('+shared')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/strumpack/package.py]: There is no shared variant.

File: var/spack/repos/builtin/packages/strumpack/package.py [hunk 4]
```diff
@@ -59,8 +64,11 @@ class Strumpack(CMakePackage):
     depends_on('parmetis', when='+parmetis')
     depends_on('scotch~metis', when='+scotch')
     depends_on('scotch~metis+mpi', when='+scotch+mpi')
+    depends_on('butterflypack@1.1.0:', when='+butterflypack+mpi')
 
     conflicts('+parmetis', when='~mpi')
+    conflicts('+butterflypack', when='~mpi')
+    conflicts('+butterflypack', when='strumpack@:3.2.0')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/strumpack/package.py]: Are we sure we want `+butterflypack` to default to True? That means building any older version will fail unless someone manually changes to `~butterflypack`.

Comment (pghysels) [reply var/spack/repos/builtin/packages/strumpack/package.py]: +butterflypack is an important new development in strumpack, so we want this to be default.
One might 'forget' to add +butterflypack, resulting in an install with limited usability. But when you forget to add ~butterflypack, you get a useful
  ".1. "+butterflypack" conflicts with "strumpack@:3.2.0"
and you can just add that missing ~butterflypack, before waiting for the install.

Is there a way to set the default value of the variant based on the version?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/strumpack/package.py]: Alright, I'm fine with this explanation.

> Is there a way to set the default value of the variant based on the version?

Not yet unfortunately.

**Pull request merged**